commit 17dd759c67f21e34f2156abcf415e1f60605a188
Author: Herbert Xu <herbert@gondor.apana.org.au>
Date:   Wed Jul 27 06:16:28 2011 -0700

    gro: Only reset frag0 when skb can be pulled
    
    Currently skb_gro_header_slow unconditionally resets frag0 and
    frag0_len.  However, when we can't pull on the skb this leaves
    the GRO fields in an inconsistent state.
    
    This patch fixes this by only resetting those fields after the
    pskb_may_pull test.
    
    Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
    Signed-off-by: David S. Miller <davem@davemloft.net>

diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index 1d92acc0..661a077 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -1649,9 +1649,12 @@ static inline int skb_gro_header_hard(struct sk_buff *skb, unsigned int hlen)
 static inline void *skb_gro_header_slow(struct sk_buff *skb, unsigned int hlen,
 					unsigned int offset)
 {
+	if (!pskb_may_pull(skb, hlen))
+		return NULL;
+
 	NAPI_GRO_CB(skb)->frag0 = NULL;
 	NAPI_GRO_CB(skb)->frag0_len = 0;
-	return pskb_may_pull(skb, hlen) ? skb->data + offset : NULL;
+	return skb->data + offset;
 }
 
 static inline void *skb_gro_mac_header(struct sk_buff *skb)
